### DISCUSSION TITLE
Functions don't have ` __get__`

### DISCUSSION THREAD

**Discussion by adriangb** [microsoft/pyright#4906] [General]
I'm confused about the following behavior:

```python
from typing import Any, Protocol

class AnyDescriptor(Protocol):
    @property
    def __get__(self) -> Any:
        ...

def foo(cls: Any) -> None:
    pass

foo.__get__  # okay
v1: AnyDescriptor = foo  # "object" is incompatible with protocol "AnyDescriptor", "__get__" is not present
v2: AnyDescriptor = classmethod(foo)  # okay
```

MyPy has no issues with this code even with `--strict`.

I'm not sure, but I thought all functions implement `__get__` and become bound instance methods if they get an instance.

Context: I was trying to type a decorator that accepts a function, staticmethod or classmethod with a specific signature:

```python
from typing import Any

def decorator(func: Any) -> Any:  # typing enforces siganture here
    return func

def f():
    pass

class Foo:
    @decorator
    @classmethod
    def f1(cls):
        pass

    @decorator
    def f2(self):
        pass

    f3 = decorator(f)

# used like
Foo.f1.__get__(None, Foo)()
Foo.f2.__get__(None, Foo)(Foo())
Foo.f3.__get__(None, Foo)()
```

The reason I was trying to use a protocol is because until recently `classmethod` & co. were not fully type annotated. While Pyright already supports the fully typed version that is now in typeshed MyPy doesn't. Using a single protocol would also simplify a lot of ugly type annotations.

**Comment by erictraut** [comment]
There is a special class defined in typeshed's `builtin.pyi` called `function`. This class doesn't exist in the runtime, but it was added for mypy to define the additional attributes that can be found on a function object. This includes the `__get__` method plus a bunch of other attributes. Use of this special class requires special casing within the type checker. I added this special-case logic a while ago for the purposes of member access expression evaluation, which explains why `foo.__get__` evaluates without an error. However, I didn't add similar special-case logic for protocol matching. That should be straightforward to add.

I'll create an issue from this discussion topic to track this.

**Reply by adriangb** [reply]
Awesome, thank you!

With respect to the actual goal of having a decorator that accepts `classmethod`s, `staticmethod`s, `functools.partialmethod`s and such while enforcing a signature on the wrapped function, does the descriptor protocol approach seem like a good idea to you?

**Reply by erictraut** [reply]
The `classmethod` and `staticmethod` decorators are highly special-cased in type checkers, so they may not do quite what you expect when used as actual decorators. But if all callables are special-cased to expose the attributes defined in `builtins.function`, your technique should work. I wouldn't recommend using this technique in a library interface, but it's probably OK for your own internal code.
